<h3>Imports</h3>

In [1]:
import duckdb
import pandas as pd
import numpy as np
import random

<h3>Reading yearly economic status data</h3>

In [3]:
df = pd.read_csv('../../data/additionalData/economicStatusMod2.csv')
df

District  Total: All usual residents aged 16 years and over  \
0          Boston                                              57584   
1    East Lindsey                                             121420   
2         Lincoln                                              87064   
3  North Kesteven                                              98017   
4   South Holland                                              78998   
5  South Kesteven                                             118011   
6    West Lindsey                                              79120   

   Economically active (excluding full-time students):In employment  \
0                                              32943                  
1                                              54579                  
2                                              43552                  
3                                              55685                  
4                                              44457                  
5                                              66909                  
6                                              41326                  

   Economically active (excluding full-time students): Unemployed  \
0                                               1411                
1                                               3241                
2                                               2299                
3                                               1728                
4                                               1758                
5                                               2598                
6                                               1754                

   Economically active and a full-time student:In employment  \
0                                                688           
1                                               1000           
2                                               3163           
3                                               1209           
4                                                777           
5                                               1286           
6                                                812           

   Economically active and a full-time student: Unemployed  \
0                                                218         
1                                                268         
2                                               1576         
3                                                356         
4                                                217         
5                                                392         
6                                                267         

   Economically inactive  
0                  22324  
1                  62332  
2                  36474  
3                  39039  
4                  31789  
5                  46826  
6                  34961

<h3>Column modifications</h3>
We create the columns:
<ul>
    <li>"Employed" which contain columns:</li>
    <ul>
        <li>Economically active (excluding full-time students):In employment</li>
        <li>Economically active and a full-time student:In employment</li>
    </ul>
    <li>"Unemployed" which contain columns:</li>
    <ul>
        <li>Economically active (excluding full-time students):Unemployed</li>
        <li>Economically active and a full-time student:Unemployed</li>
    </ul>
</ul>

We kept the Economically Inactive column as it is.

In [4]:
cols = [s for s in df.columns.to_list() if 'In employment' in s]

df['Employed'] = df[cols[0]] + df[cols[1]]

cols = [s for s in df.columns.to_list() if 'Unemployed' in s]

df['Unemployed'] = df[cols[0]] + df[cols[1]]

<h3>Keeping relevant columns</h3>

In [5]:
df = df[['District', 'Economically inactive', 'Employed', 'Unemployed']]


<h3>Retrieving months in crimesPrices table</h3>

In [6]:
con = duckdb.connect("../../data/exploitation/crimesPrices.db")
months = con.execute("SELECT Distinct Month FROM crimesPrices").df()
con.close()

In [6]:
df

District  Economically inactive  Employed  Unemployed
0          Boston                  22324     33631        1629
1    East Lindsey                  62332     55579        3509
2         Lincoln                  36474     46715        3875
3  North Kesteven                  39039     56894        2084
4   South Holland                  31789     45234        1975
5  South Kesteven                  46826     68195        2990
6    West Lindsey                  34961     42138        2021

<h3>Generating data</h3>
For each district we generate n values per district for Economically inactive, Employed and Unemployed where n is the number of different months in the crimesPrices table.

When generating the n values for a district, we take the original value and apply a random error between ±5%.

For example, say we want to calculate the n values for Employment in the district Boston. First, we would take the original value for Employment in Boston, 33631.

Next, for the n months, we apply a random error between ±5%. So, let's say that for the month 2022-05 we get a random value of 1.02 = 1 + 0.02 (the random error). The calculated value would be 33631*1.02 = 34303.62. Finally, as we cannot have 0.62 people, we floor the number. So the final value would be 34303.

We do this for all districts and each of the columns Economically inactive, Employed and Unemployed.

In [7]:
finaldf = pd.DataFrame(columns= ['Month'] + list(df.columns))
for row in df.values:
    for month in months.values:
        finaldf = finaldf.append(pd.DataFrame([list(month) + list(row)], columns=['Month'] + list(df.columns)), ignore_index=True)

for att in finaldf.columns[2:]:
    finaldf[att] = (finaldf[att] * pd.DataFrame(np.random.uniform(0.95, 1.05, size=(len(finaldf.values), 1)))[0]).apply(np.floor)

finaldf

Month      District Economically inactive Employed Unemployed
0   2022-05        Boston                 22359    34826       1555
1   2022-04        Boston                 22295    33244       1579
2   2022-03        Boston                 23062    33775       1558
3   2022-02        Boston                 21774    34109       1643
4   2022-01        Boston                 21467    34849       1665
..      ...           ...                   ...      ...        ...
79  2021-10  West Lindsey                 33279    43686       2007
80  2021-09  West Lindsey                 34555    43732       2081
81  2021-08  West Lindsey                 36189    41255       2033
82  2021-07  West Lindsey                 36393    41119       1961
83  2021-06  West Lindsey                 34827    42915       2118

[84 rows x 5 columns]

In [8]:
finaldf.to_csv('../../data/landing/temporal/economicStatus.csv', index=False)